### ISSUE TITLE
Invalid optimization for ((int[])(object)uintArray) is uint[]

### ISSUE THREAD

**Issue by RamType0** [dotnet/roslyn#53240]
**Version Used**: 
Microsoft Visual Studio Community 2019
Version 16.8.5
**Steps to Reproduce**:

``` C#
using System;
public class C {
    public void M() {
        uint[] uintArray = new uint[1];
        object uintArrayAsObject = uintArray;
        int[] uintArrayAsIntArray = (int[])uintArrayAsObject;
        Console.WriteLine(uintArray is uint[]);
        Console.WriteLine(uintArrayAsObject is uint[]);
        Console.WriteLine(uintArrayAsIntArray is uint[]);
    }
}
```

**Expected Behavior**:
Console says 
> True
True
True

**Actual Behavior**:
Console says 
> True
True
False

**Note**:
Current C# compiler treats `uintArray is uint[]` as `uintArray is not null`,
treats `uintArrayAsObject is uint[]` as `uintArray is uint[]`,
treats `uintArrayAsIntArray is uint[]` as `false`.

Same issue for other array of same-sized integer expected.

**Comment by CyrusNajmabadi** [comment]
I think this is by design (@jaredpar to confirm).  `uintArray is uint[]` and `uintArrayAsIntArray is uint[]` are tests against the static known type of the value.  So in the former, it's just a non-null test as the type isn't already known to be a `uint[]`.  the latter is just always false as *from a static perspective* a `uint[]` is not an `int[]`.  

The middle case is the interesting one.  From a static perspective, the type is not known.  So this becomes a dynamic check to the runtime.  In this case the runtime is is interesting and treats these types similarly.  So even though the language does not consider these types the same, the runtime does.  So you end up with this unexpected, by actually expected result :)

**Comment by RamType0** [comment]
The fundamental issue is that Roslyn think that uint[] is not int[], and it is inconsistent between runtime's policy.

**Comment by CyrusNajmabadi** [comment]
>The fundamental issue is that Roslyn think that uint[] is not int[]

Roslyn follows the C# language specification.  I'll read it again, but i don't think these would be the same statically as far as C# is concerned.

**Comment by CyrusNajmabadi** [comment]
Yes.  Looking at the spec, the only implicit (or explicit) conversion for arrays is:

```
From an array_type S with an element type SE to an array_type T with an element type TE, provided all of the following are true:
S and T differ only in element type. In other words, S and T have the same number of dimensions.
Both SE and TE are reference_types.
An implicit reference conversion exists from SE to TE.
```

As SE and TE are not reference types, there is no conversion (implicit or explicit) between these two arrays.  So the compiler is doing things properly.

**Comment by RamType0** [comment]
If the compiler correctly follows C# language specification,

```C#
int[] uintArrayAsIntArray = (int[])uintArrayAsObject;
```
This should fail.
So the compiler doesn't follow any of C# language specification or .NET Runtime policy.

**Comment by CyrusNajmabadi** [comment]
> This should fail.

This is legal as per the c# language. See the conversions section of the spec

**Comment by RamType0** [comment]
> > This should fail.
> 
> This is legal as per the c# language. See the conversions section of the spec

I mean it fail at runtime(not at compile time).

**Comment by CyrusNajmabadi** [comment]
> I mean it fail at runtime(not at compile time).

That would be an issue for dotnet/runtime.  The c# compiler is properly implementing the c# language specification and emitting the right IL.  Note that it would be unlikely for the runtime to change this as it would almost certainly break some customers.

**Comment by RamType0** [comment]
> The c# compiler is properly implementing the c# language specification and emitting the right IL.

It is right as far as explicit conversion in C# and the IL instruction `castclass` is same.
But it is not same for casting uint[] instance to int[].

Moreover, [`uintArrayAsIntArray is uint[]` pattern matches "D and T are the same type"](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#the-is-operator).
C# language specification treats it as true.
